src/test/codeactions.test.ts

```diff
@@ -0,0 +1,279 @@
+  const ide = new IdeAnalyzer();
+  ide.setOpenFileContents(options.rig.resolve('package.json'), contents);
+  const actions = await ide.getCodeActions(
+    options.rig.resolve('package.json'),
+    OffsetToPositionConverter.createUncachedForTest(contents).toIdeRange({
+      offset,
+      length: 0,
+    })
+  );
+  return {contents, actions};
+}
+
+async function assertCodeAction(options: {
+  rig: WireitTestRig;
+  contentsWithPipe: string | object;
+  expectedOutput: string | object;
+  expectedTitle: string;
+}) {
+  const {contentsWithPipe, expectedOutput} = options;
+  if (
+    typeof contentsWithPipe === 'object' &&
+    typeof expectedOutput === 'object'
+  ) {
+    // Verify that we behave correctly even in the face of different formatting.
+    await assertCodeAction({
+      ...options,
+      contentsWithPipe: JSON.stringify(contentsWithPipe, null, 2),
+    });
+    await assertCodeAction({
+      ...options,
+      contentsWithPipe: JSON.stringify(contentsWithPipe, null, 8),
+    });
+    await assertCodeAction({
+      ...options,
+      contentsWithPipe: JSON.stringify(contentsWithPipe, null, '\t'),
+    });
+  }
+  const {actions, contents} = await getCodeActions(options);
+  assert.equal(
+    actions.map((a) => a.title),
+    [options.expectedTitle],
+    `Error getting code actions for ${JSON.stringify(options.contentsWithPipe)}`
+  );
+  const [action] = actions;
+  const newContents = applyEdit(options.rig, contents, action);
+  if (typeof options.expectedOutput === 'string') {
+    assert.equal(newContents, options.expectedOutput);
+  } else {
+    assert.equal(JSON.parse(newContents), options.expectedOutput);
+  }
+}
+
+async function assertNoCodeActions(options: {
+  rig: WireitTestRig;
+  contentsWithPipe: string | object;
+}) {
+  const {actions} = await getCodeActions(options);
+  assert.equal(
+    actions.map((a) => a.title),
+    [],
+    `Expected no code action in: ${JSON.stringify(options.contentsWithPipe)}\n`
+  );
+}
+
+function applyEdit(
+  rig: WireitTestRig,
+  before: string,
+  action: CodeAction
+): string {
+  if (action.edit == null) {
+    throw new Error(`Action ${action.title} had no edit`);
+  }
+  const edit = action.edit;
+  assert.equal(Object.keys(edit), ['changes']);
+  const filename = rig.resolve('package.json');
+  assert.equal(Object.keys(edit?.changes ?? {}), [filename]);
+  const textEdits = edit?.changes?.[filename];
+  if (textEdits == null) {
+    throw new Error(`Action ${action.title} had no edits for ${filename}`);
+  }
+  const converter = OffsetToPositionConverter.createUncachedForTest(before);
+  return jsonParser.applyEdits(
+    before,
+    textEdits.map((e): jsonParser.Edit => {
+      return {
+        content: e.newText,
+        ...converter.ideRangeToRange(e.range),
+      };
+    })
+  );
+}
+
+test('can refactor a script to use wireit', async ({rig}) => {
+  await assertCodeAction({
+    rig,
+    contentsWithPipe: {scripts: {test: "echo| 'test'"}},
```

Comment:
What is the significance of the pipe characters in this test file? I think I missed something.

Reply:
Ah yeah, added a comment. That's to indicate the location of the carat, where the user's text input cursor is when they request code actions.

Reply:
Oh I see, makes sense thanks.